aten/src/ATen/native/cuda/FractionalMaxPool3d.cu

```diff
@@ -0,0 +1,410 @@
+      gradOutput[batch][plane][outputH][outputW][outputT]
+      );
+    }
+  }
+
+void fractional_max_pool3d_out_cuda_template(
+  Tensor& output,
+  Tensor& indices,
+  const Tensor& input,
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& randomSamples) {
+    int planeDim = 0;
+    int dimh = 1;
+    int dimw = 2;
+    int dimt = 3;
+    int numBatch = 1;
+
+    int outputT = output_size[2];
+    int outputW = output_size[1];
+    int outputH = output_size[0];
+    int poolSizeT = pool_size[2];
+    int poolSizeW = pool_size[1];
+    int poolSizeH = pool_size[0];
+
+    int64_t ndims = input.ndimension();
+    AT_CHECK(
+      input.numel() != 0 && (ndims == 4 || ndims == 5),
+      "fractional_max_pool3d_out_cuda_template(): ",
+      "non-empty 4D or 5D (batch mode) tensor expected for input, but got: ",
+      ndims);
+
+    if (ndims == 5) {
+      numBatch = input.size(0);
+      planeDim++;
+      dimh++;
+      dimw++;
+      dimt++;
+    }
+
+    /* sizes */
+    int64_t numPlanes = input.size(planeDim);
+    int64_t inputH = input.size(dimh);
+    int64_t inputW = input.size(dimw);
+    int64_t inputT = input.size(dimt);
+
+    AT_CHECK(
+      outputH + poolSizeH - 1 < inputH,
+      "fractional_max_pool3d_out_cuda_template(): ",
+      "pool height (", poolSizeH, ") too large relative to input height (",
+      inputH, ")");
+    AT_CHECK(
+      outputW + poolSizeW - 1 < inputW,
+      "fractional_max_pool3d_out_cuda_template(): ",
+      "pool width (", poolSizeW, ") too large relative to input width (",
+      inputW, ")");
+    AT_CHECK(
+      outputT + poolSizeT - 1 < inputT,
+      "fractional_max_pool3d_out_cuda_template(): ",
+      "pool time (", poolSizeT, ") too large relative to input time (",
+      inputT, ")");
+
+    if (ndims == 4) {
+      /* resize output */
+      output.resize_({numPlanes, outputH, outputW, outputT});
+      /* indices will contain the locations for each output point */
+      indices.resize_({numPlanes, outputH, outputW, outputT});
+    } else {
+      /* resize output */
+      output.resize_({numBatch, numPlanes, outputH, outputW, outputT});
+      /* indices will contain the locations for each output point */
+      indices.resize_({numBatch, numPlanes, outputH, outputW, outputT});
+    }
+
+    auto output_ = output;
+    auto indices_ = indices;
+    auto input_ = input;
+    if(ndims == 4) {
+      output_.reshape({1, numPlanes, outputH, outputW, outputT});
+      indices_.reshape({1, numPlanes, outputH, outputW, outputT});
+      input_.reshape({1, numPlanes, inputH, inputW, inputT});
+    }
+
+    // block is limited to 4 warps
+    // grid handles overflow per each plane
+    int outputPlaneSize = output_.size(2) *
+      output_.size(3) * output_.size(4);
+    dim3 grid(
+      (outputPlaneSize + 127) / 128,
+      input_.size(1),
+      input_.size(0));
+    dim3 block(outputPlaneSize > 128 ? 128 : outputPlaneSize);
+
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
```

Comment:
It looks like we lost the specializations for pool size 2-7, although I am not sure we actually care.

Reply:
It is still preserved in the fractional_max_pool3d_out_frame function.